crypto/s2n_tls13_keys.c

```diff
@@ -172,14 +172,19 @@ int s2n_tls13_derive_binder_key(struct s2n_tls13_keys *keys, struct s2n_psk *psk
 /*
  * Derives early secrets
  */
-int s2n_tls13_derive_early_secrets(struct s2n_tls13_keys *keys)
+int s2n_tls13_derive_early_secrets(struct s2n_tls13_keys *keys, struct s2n_psk *psk)
 {
     notnull_check(keys);
 
-    s2n_tls13_key_blob(psk_ikm, keys->size); /* in 1-RTT, PSK is 0-filled of key length */
+    if (psk == NULL) {
+        /* in 1-RTT, PSK is 0-filled of key length */
+        s2n_tls13_key_blob(psk_ikm, keys->size);
 
-    /* Early Secret */
-    GUARD(s2n_hkdf_extract(&keys->hmac, keys->hmac_algorithm, &zero_length_blob, &psk_ikm, &keys->extract_secret));
+        /* Early Secret */
+        GUARD(s2n_hkdf_extract(&keys->hmac, keys->hmac_algorithm, &zero_length_blob, &psk_ikm, &keys->extract_secret));
+    } else {
+        keys->extract_secret = psk->early_secret;
```

Comment:
That won't actually be sufficient to make sure that the early_secret exists. It's a blob structure, not a pointer. We could check that the blob isn't empty tho.

Reply:
Is this necessary? I'm not sure what error we'd throw here.

Reply:
It's not necessary as in "a likely failure case we'll need to handle." It should be impossible for the early secret to be empty here. The check would just be a sanity check to verify assumptions, and it could just throw a safety error.

You could write a test to verify an empty secret fails. If we error later in the calculation if the secret is empty, the check here would be unnecessary. But if later crypto methods will happily accept an empty secret, we should add the "not empty" checks.